### PR TITLE
DX: add route context to the dynamic errors

### PR DESCRIPTION
**Pull request by huozhi**
### What 

Given user infomation when the dynamic errors are thrown, e.g. bad `cookies` or `headers` usages. Now users can tell through the error information to see which pathname is broken, and trace down the usage.

#### before

```
Page couldn't be rendered statically because ...
This page needs to bail out of prerendering at this point because ...
```

#### after

```
Route /cookies couldn't be rendered statically because ...
Route /server needs to bail out of prerendering at this point because ...
```

### Why

When you have multi pages in your app, such as 100+, and many page might uses these. This is hard to trace down where exactly the error is from

Closes NEXT-2283
Cloese NEXT-2265

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 71e8bf082a87336555b9761a9fadd6122dc4cc64

`pnpm test-dev test/development/acceptance-app/dynamic-error.test.ts`
- dynamic = "error" in devmode > should show error overlay when dynamic is forced

<details>
<summary>Expand output</summary>

  ● dynamic = "error" in devmode › should show error overlay when dynamic is forced

    expect(received).toMatchInlineSnapshot(snapshot)

    Snapshot name: `dynamic = "error" in devmode should show error overlay when dynamic is forced 1`

    - Snapshot  - 1
    + Received  + 1

    - "Error: Route /server with `dynamic = "error"` couldn't be rendered statically because it used `cookies`. See more info here: https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic#dynamic-rendering"
    + "Error: Route /server?_rsc=1sucx with `dynamic = "error"` couldn't be rendered statically because it used `cookies`. See more info here: https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic#dynamic-rendering"

      33 |     await session.hasRedbox()
      34 |     console.log(await session.getRedboxDescription())
    > 35 |     expect(await session.getRedboxDescription()).toMatchInlineSnapshot(
         |                                                  ^
      36 |       `"Error: Route /server with \`dynamic = "error"\` couldn't be rendered statically because it used \`cookies\`. See more info here: https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic#dynamic-rendering"`
      37 |     )
      38 |

      at Object.toMatchInlineSnapshot (development/acceptance-app/dynamic-error.test.ts:35:50)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js fix/add-missing-context-for-rendering | Change |
| - | - | - | - |
| buildDuration | 11.7s | 11.6s | N/A |
| buildDurationCached | 6s | 4.9s | N/A |
| nodeModulesSize | 200 MB | 200 MB | ⚠️ +6.23 kB |
| nextStartRea..uration (ms) | 435ms | 433ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/add-missing-context-for-rendering | Change |
| - | - | - | - |
| 3f784ff6-HASH.js gzip | 53.4 kB | 53.4 kB | ✓ |
| 423.HASH.js gzip | 185 B | 181 B | N/A |
| 68-HASH.js gzip | 29.8 kB | 29.8 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 240 B | 240 B | ✓ |
| main-HASH.js gzip | 31.8 kB | 31.8 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 101 kB | 101 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/add-missing-context-for-rendering | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/add-missing-context-for-rendering | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 502 B | 502 B | ✓ |
| css-HASH.js gzip | 320 B | 322 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 256 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 368 B | 369 B | N/A |
| image-HASH.js gzip | 4.18 kB | 4.18 kB | N/A |
| index-HASH.js gzip | 257 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.61 kB | N/A |
| routerDirect..HASH.js gzip | 310 B | 311 B | N/A |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 306 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 608 B | 608 B | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/add-missing-context-for-rendering | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 484 B | ✓ |
| Overall change | 484 B | 484 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/add-missing-context-for-rendering | Change |
| - | - | - | - |
| index.html gzip | 529 B | 527 B | N/A |
| link.html gzip | 541 B | 540 B | N/A |
| withRouter.html gzip | 523 B | 522 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/add-missing-context-for-rendering | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94 kB | 94 kB | N/A |
| page.js gzip | 150 kB | 150 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/add-missing-context-for-rendering | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 622 B | 623 B | N/A |
| middleware-r..fest.js gzip | 151 B | 149 B | N/A |
| middleware.js gzip | 37.6 kB | 37.6 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 1.92 kB | 1.92 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js fix/add-missing-context-for-rendering | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 170 kB | 170 kB | ✓ |
| app-page-exp..prod.js gzip | 95.7 kB | 95.7 kB | ✓ |
| app-page-tur..prod.js gzip | 96.4 kB | 96.4 kB | ✓ |
| app-page-tur..prod.js gzip | 91 kB | 91 kB | ✓ |
| app-page.run...dev.js gzip | 142 kB | 142 kB | ✓ |
| app-page.run..prod.js gzip | 90.3 kB | 90.3 kB | ✓ |
| app-route-ex...dev.js gzip | 22.2 kB | 22.2 kB | N/A |
| app-route-ex..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route-tu..prod.js gzip | 14.5 kB | 14.5 kB | ✓ |
| app-route.ru...dev.js gzip | 21.6 kB | 21.6 kB | N/A |
| app-route.ru..prod.js gzip | 14.5 kB | 14.5 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.42 kB | 9.42 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.69 kB | 9.69 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.42 kB | 9.42 kB | ✓ |
| pages-turbo...prod.js gzip | 22 kB | 22 kB | ✓ |
| pages.runtim...dev.js gzip | 22.6 kB | 22.6 kB | ✓ |
| pages.runtim..prod.js gzip | 22 kB | 22 kB | ✓ |
| server.runti..prod.js gzip | 49.7 kB | 49.7 kB | ✓ |
| Overall change | 889 kB | 889 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>68-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..ntime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-ex..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route-tu..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-route.ru..time.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 71e8bf082a87336555b9761a9fadd6122dc4cc64

File: packages/next/src/server/app-render/dynamic-rendering.ts [hunk 1]
```diff
@@ -40,14 +40,15 @@ export function markCurrentScopeAsDynamic(
   store: StaticGenerationStore,
   expression: string
 ): void {
+  const pathname = getPathname(store.urlPathname)
```
Comment (ztanner) [inline packages/next/src/server/app-render/dynamic-rendering.ts:43]: Is this the same as `store.pagePath`? (I can never remember what that is)

Comment (huozhi) [reply packages/next/src/server/app-render/dynamic-rendering.ts:43]: pagePath is the page route, for app router it's `/pathname/page` for pages and `/pathname/route` for custom routes, and the type was string | undefined. So url path is pretty safe to go that I only need to remove the rsc query

**Review by ijjk — APPROVED** [review]

**Review by ztanner — APPROVED** [review]

**Pull request merged**